### ISSUE TITLE
Remove meta data from graph nodes?

### ISSUE THREAD

**Issue by MichaelClerx** [ModellingWebLab/cellmlmanip#149]
After #142 , we'll have symbols (variables) represented by `VariableDummy` objects that contain meta data for that variable.
In addition, after parsing etc. is fully finished we have a graph where each node represents a variable, and has a copy of the meta data from the symbols.

Following a comment from @jonc125 on #148, this is a proposal to remove the meta data from the graph. ~Instead, you'll be able to get it from e.g. `node.symbol.name`, or `node.symbol.units`~ At the moment, if you iterate over the graph nodes you already get the symbols (VariableDummy objects), so you can access the meta data from them

Thoughts @MauriceHendrix @skeating ?

**Comment by MichaelClerx** [comment]
The benefit of this would be that we lose some code where the meta data is copied, and remove the possibility that the two go out of sync. (For example, if you change the meta data in the graph it currently doesn't affect the symbol, or vice versa)

**Comment by MauriceHendrix** [comment]
Nice idea, I'm assuming things lik summarize_units will disappear?
One thing to be aware of is the way you currently need to check if 2 units are the same. It might be worth moving that same_units check method into the VariableDummy for convenience.

Finally we'd need to decide what methids such as get_equations_for take as input (symbol or the new VariableDummy

**Comment by MichaelClerx** [comment]
> Nice idea, I'm assuming things lik summarize_units will disappear?

Don't quite understand this bit. Summarise_units acts on a sympy expression, e.g. an RHS, so don't see how this would be affected?

**Comment by MichaelClerx** [comment]
We could certainly move some methods into the VariableDummy, if you like, but think that'd be a separate issue to this one

**Comment by MichaelClerx** [comment]
> Finally we'd need to decide what methids such as get_equations_for take as input (symbol or the new VariableDummy

The `VariableDummy` _is_ a symbol! That's the nice thing about it :-)
Instead of symbols + dummies + graph nodes we now have symbols(=dummies) + graph nodes

**Comment by jonc125** [comment]
> We could certainly move some methods into the VariableDummy, if you like, but think that'd be a separate issue to this one

I'm not sure I'd want to do it unless there was a very strong reason, either. What we're doing isn't recommended by sympy after all!

**Comment by MauriceHendrix** [comment]
> > We could certainly move some methods into the VariableDummy, if you like, but think that'd be a separate issue to this one
> 
> I'm not sure I'd want to do it unless there was a very strong reason, either. What we're doing isn't recommended by sympy after all!

Then why are we doing it? Didn't we try this before? I seem to remember we abandoned the idea after some experiments a while back. I do agree it would be nice, but if it's not recommended by SymPy maybe we should stop to think why.

**Comment by MichaelClerx** [comment]
See #142 and #148 for that bit of the discussion :-)

**Comment by MauriceHendrix** [comment]
Hhm well let's hope it works properly. If it's just because the authors didn't anticipate it then fine, but if the internals do some less then robust type checks it might be an issue.

**Comment by MichaelClerx** [comment]
Hmmmm there's an issue with this idea (not having the meta data on the graph) and I'm not sure if/how to resolve it.

The problem is that the Model class has two representations of everything: (1) the symbols (VariableDummy) objects, and a list of equations (which I suppose could be a mapping from LHS to RHS expressions instead), and (2) a graph

Some properties, e.g. variable type, are only set when creating the graph. There's no mechanism to unset them if the cached graph is deleted. There's also no nice way for the user to know if they can trust the meta info inside the symbols. (They have to access the graph before it's guaranteed they're set, but that's a very strange condition!)

I think it all goes back to the Model class having 2 functions: (1) containing partial information for use during parsing and (2) maintaining an internally consistent representation that users can access

The graph is pretty much part 2

**Comment by MauriceHendrix** [comment]
> > Nice idea, I'm assuming things lik summarize_units will disappear?
> 
> Don't quite understand this bit. Summarise_units acts on a sympy expression, e.g. an RHS, so don't see how this would be affected?

Not really. I was always doing it on LHS as doing on RHS never worked properly!

**Comment by MichaelClerx** [comment]
I still don't understand how that's related to this issue! If you want to get the units of a symbol then yes you can use symbol.units now. But doing that on the RHS should check if the RHS equation evaluates to something in the same units, so quite a different use case

**Comment by MichaelClerx** [comment]
> Hmmmm there's an issue with this idea (not having the meta data on the graph) and I'm not sure if/how to resolve it.
> 
> The problem is that the Model class has two representations of everything: (1) the symbols (VariableDummy) objects, and a list of equations (which I suppose could be a mapping from LHS to RHS expressions instead), and (2) a graph
> 
> Some properties, e.g. variable type, are only set when creating the graph. There's no mechanism to unset them if the cached graph is deleted. There's also no nice way for the user to know if they can trust the meta info inside the symbols. (They have to access the graph before it's guaranteed they're set, but that's a very strange condition!)
> 
> I think it all goes back to the Model class having 2 functions: (1) containing partial information for use during parsing and (2) maintaining an internally consistent representation that users can access (see #120 )
> 
> The graph is pretty much part 2

One way to resolve it would be to do the exact opposite of what this issue says, and make sure that users know _only_ to get meta data from the graph. For example we can make all the VariableDummy properties hidden/private. In addition we'd need to add a bunch of explanation to the model docstring

**Comment by MauriceHendrix** [comment]
> I still don't understand how that's related to this issue! If you want to get the units of a symbol 

Just trying to explain where the question was coming from: I thought summarize_units was the way to get units for a symbol as that's what I'm using now. Last time I checked it didn't work on the RHS (but that might have been fixed)

**Comment by jonc125** [comment]
> Hmmmm there's an issue with this idea (not having the meta data on the graph) and I'm not sure if/how to resolve it.
> 
> The problem is that the Model class has two representations of everything: (1) the symbols (VariableDummy) objects, and a list of equations (which I suppose could be a mapping from LHS to RHS expressions instead), and (2) a graph
> 
> Some properties, e.g. variable type, are only set when creating the graph. There's no mechanism to unset them if the cached graph is deleted. There's also no nice way for the user to know if they can trust the meta info inside the symbols. (They have to access the graph before it's guaranteed they're set, but that's a very strange condition!)
> 
> I think it all goes back to the Model class having 2 functions: (1) containing partial information for use during parsing and (2) maintaining an internally consistent representation that users can access (see #120 )
> 
> The graph is pretty much part 2

In my view, the list of equations is an internal detail for storing maths. (And perhaps a set would make more sense, as the ordering is arbitrary, and we have operations like add/remove? But anyway, separate discussion!) When you're manipulating the model, you're using functions that update this internal representation, and the associated metadata about variables & units. The manipulation functions should ensure the metadata is as consistent as necessary for this purpose.

Once you have the model in a final state, the graph representation comes into play, and is intended primarily for tracking dependencies between equations, so you can do topological sorting and filtering according to what equations are needed in particular contexts.

**Comment by skeating** [comment]
Yes it would be ideal to only ever have information in one place and construct the graph such that any queries refer to internal structure. But then we do have the issue of them potentially going out of sync.

It seems that alot of queries use the graph object and so if someone is making changes we need to be sure that we are keeping things consistent.

One solution would be to use an Observer https://refactoring.guru/design-patterns/observer/python/example to track when something in the internal structure is changed and then when calling the graph it notes things have changed and regenerates rather than use the cached graph.

**Comment by jonc125** [comment]
I think Michael's the use of `@property` and modification methods invalidating the graph should have the same effect as the observer pattern, with less code. Would then only become an issue if a user keeps their own copy of the graph - but we can document to not do that!

**Comment by MichaelClerx** [comment]
Now thinking to 

1. Remove `type` from the `VariableDummy` objects, as it derives from the equations not the variable
2. Remove all but `type` from the graph
3. Add some documentation / example code to the `Model` docstring

@MauriceHendrix would point 2 make life harder for you? Not entirely necessary to do pt 2, but it would be cleaner I think

**Comment by MauriceHendrix** [comment]
I don't use the graph directly so as long as the API methods keep working properly (and the sorted ones return the same order) it should be ok I think.

I can always just actually test it with my code to see the impact once you have a PR
